### DISCUSSION TITLE
Setting up backpane with Aspire + Azure Redis + Entra Auth

### DISCUSSION THREAD

**Discussion by onionhammer** [ZiggyCreatures/FusionCache#256] [General]
I've written the following extension methods to try to connect together the components for FusionCache to utilize the same connection as the distributed redis cache

The first is to add an IConnectionMultiplexer which utilizes the [ConfigureForAzureWithTokenCredentialAsync](https://github.com/Azure/Microsoft.Azure.StackExchangeRedis)

The second is meant to add the fusion cache and utilize the IConnectionMultiplexer

First question: Is this sound? Am I missing something here?

If not, I'm seeing the following error from the backplane in the logs; things seem to work fine locally, however.. 

"The message timed out in the backlog attempting to send because no connection became available (5000ms) - Last Connection Exception: AuthenticationFailure"

The redis distributed cache layer appears to work but the backplane doesnt

### Code

```C#
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using StackExchange.Redis;

public static class RedisExtensions
{
    public static IHostApplicationBuilder AddMyRedisContext(
        this IHostApplicationBuilder builder, string name)
    {
        var credentials = new DefaultAzureCredential(...);

        builder.AddRedisClient(name, configureOptions: options =>
        {
            // If not localhost & no password, use Azure credentials
            if (string.IsNullOrWhiteSpace(options.Password) && options.SslHost?.Contains("localhost") == false)
                options.ConfigureForAzureWithTokenCredentialAsync(credentials).Wait();
        });

        builder.Services.AddKeyedSingleton(name, (sp, fact) =>
            // Get keyed client
            sp.GetRequiredService<IConnectionMultiplexer>()
        );

        return builder;
    }
}
```

```C#
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using ZiggyCreatures.Caching.Fusion;
using ZiggyCreatures.Caching.Fusion.Backplane.StackExchangeRedis;
using ZiggyCreatures.Caching.Fusion.Serialization.SystemTextJson;

public static class CacheExtensions
{
    public static IHostApplicationBuilder AddMyFusionCache(this IHostApplicationBuilder builder, string name)
    {
        var cacheBulder = builder.Services.AddFusionCache();

        builder.AddRedisDistributedCache(name);

        cacheBulder
            .WithSerializer(new FusionCacheSystemTextJsonSerializer());

        cacheBulder
            .WithRegisteredDistributedCache();

        cacheBulder.WithBackplane(sp =>
        {
            var logger      = sp.GetRequiredService<ILogger<RedisBackplane>>();
            var multiplexer = sp.GetRequiredService<IConnectionMultiplexer>();

            var options = new RedisBackplaneOptions()
            {
                Configuration = multiplexer.Configuration
            };

            var result = new RedisBackplane(options, logger);

            return result;
        });

        return builder;
    }
}
```

**Comment by onionhammer** [comment]
Ah nevermind, I see now.. Configuration is just a string. I guess the real question is - so it looks like something like this might work:

```C#
            cacheBulder.WithBackplane(sp =>
            {
                var logger      = sp.GetRequiredService<ILogger<RedisBackplane>>();
                var multiplexer = sp.GetRequiredService<IConnectionMultiplexer>();

                var options = new RedisBackplaneOptions()
                {
                    ConnectionMultiplexerFactory = () => Task.FromResult(multiplexer),
                };

                var result = new RedisBackplane(options, logger);

                return result;
            });
```

But I would love recommendations on the 'proper' way to do this

**Reply by jodydonetti** [reply]
Hi @onionhammer , yep you got it: you should use the [ConnectionMultiplexerFactory](https://github.com/ZiggyCreatures/FusionCache/blob/main/src/ZiggyCreatures.FusionCache.Backplane.StackExchangeRedis/RedisBackplaneOptions.cs#L29), so you can define a factory and pass the same one to both the distributed cache and the backplane.

Hope this helps.

**Reply by jodydonetti** [reply]
Oh but wait, I fixated too much on the connection multiplexer only, but there's more to it: we are talking about what the whole "Aspire + FusionCache experience" can look like!

[Here](https://github.com/ZiggyCreatures/FusionCache/issues/237) I am exploring exactly this, taking suggestions, designing it, etc.

I'd like to know more!

**Reply by onionhammer** [reply]
Unfortunately Aspire doesn't seem to utilize entra creds, I believe it just pulls the redis key into a keyvault secret, then uses that.. I'm going a bit beyond that

**Reply by jodydonetti** [reply]
Hi @onionhammer , it seems strange that Aspire launched in GA without proper support for Azure creds, I'll try to look into that (eg: I'm looking at [this](https://github.com/dotnet/aspire/blob/d215c528c07c7919c3ac30b35d92f4e51a60523b/src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs) right now)...

Anyway, although maybe not "beautifully", but is this working now?

**Reply by jodydonetti** [reply]
ps: in general though, you can use the Redis backplane by passing anything from a string `Configuration`, a `ConfigurationOptions` object or even directly a `ConnectionMultiplexerFactory` and they, in turn, can be created manually, required via "normal" DI or even keyed DI, so I think w can make this work.

To make it work more "natively" instead, as I said that is what I'm looking into in #237 so I'd be glad to have you input there, if you like 😬

**Reply by onionhammer** [reply]
Yes! it is working now, thanks! I'll add some feedback on 237

Not sure why they didn't go the extra mile and add entra auth to azure postgres and azure redis components... a question for their team I guess.

**Reply by onionhammer** [reply]
This isn't really relevant to Aspire, but this would be nice to have an extension method that does this, and then docs that guide users down that path

```C#
builder.AddRedisDistributedCache(name); // <-- Aspire

cacheBulder
    .WithSerializer(new FusionCacheSystemTextJsonSerializer())
    .WithRegisteredDistributedCache()
    .WithRegisteredStackExchangeRedisBackplane(); // <-- Use registered IConnectionMultiplexer
```

Beyond that, I think Aspire should pretty much just work without you having to do much special, beyond documenting common use-cases well; perhaps recommend putting an extension method to configure everything in the ServiceDefaults project

**Reply by jodydonetti** [reply]
Hi @onionhammer , that is the idea I'd like go with: some new ext methods to ease the setup.

Will look into this more, and I'm definitely open to ideas.

Thanks!

**Reply by fuad-bassam** [reply]
HI @jodydonetti I work with .net 9 Aspire and I'm trying to add a  [**backplane**](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/Backplane.md), but I can't find a way to register a component in the DI using the Reference for the Redis container. I'm unsure how to register it or test it. I looked into this [FEATURE 🔺 Add .NET Aspire support #237](https://github.com/ZiggyCreatures/FusionCache/issues/237) and this [DependencyInjection](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/DependencyInjection.md) but it didn't work for me. could you please show me what I'm missing or some documents related to it,

this is my current implementation 
```
var builder = DistributedApplication.CreateBuilder(args);
var redis = builder.AddRedis("rediscache").WithRedisInsight();

builder.AddProject<Projects.net9CachingPOC>("net9cachingpoc")
       .WithReference(redis)
       
       .WaitFor(redis)
       .WithReplicas(2);
builder.Build().Run();
```

```
var builder = WebApplication.CreateBuilder(args);

builder.AddServiceDefaults();
// Add services to the container.


builder.Services.AddControllers().AddApplicationPart(typeof(Presentation.AssemblyReferenceForWebProject).Assembly);

// Learn more about configuring OpenAPI at https://aka.ms/aspnet/openapi
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

builder.AddRedisDistributedCache("rediscache");


var x = builder.Services.AddFusionCache()
    .WithDefaultEntryOptions(new FusionCacheEntryOptions
    {
        Duration = TimeSpan.FromSeconds(2),
        Priority = CacheItemPriority.Low,
        IsFailSafeEnabled = true,
        FailSafeMaxDuration = TimeSpan.FromHours(1),
        FailSafeThrottleDuration = TimeSpan.FromSeconds(1),

        FactorySoftTimeout = TimeSpan.FromMilliseconds(200),
        FactoryHardTimeout = TimeSpan.FromMilliseconds(1500),

        DistributedCacheSoftTimeout = TimeSpan.FromMilliseconds(1500),
        DistributedCacheHardTimeout = TimeSpan.FromMilliseconds(3000),
        AllowBackgroundDistributedCacheOperations = true,

    }).WithSerializer(new FusionCacheNewtonsoftJsonSerializer()
).WithRegisteredDistributedCache();
```

**Reply by jodydonetti** [reply]
Hi @fuad-bassam sorry for the delay, I somehow missed this.
Have you been maybe able to solve this issue?
